stdlib/http/src/main/java/org/ballerinalang/net/http/websocket/WebSocketUtil.java

```diff
@@ -224,54 +246,160 @@ private static ErrorValue createErrorCause(String message, String reason, BPacka
+            createDelay(failoverContext.getFailoverInterval());
+            establishWebSocketConnection(createWebSocketClientConnector(targets.get(currentIndex).toString(),
+                       webSocketClient), webSocketClient, wsService);
+            return true;
+        }
+        if (logger.isDebugEnabled()) {
+            logger.debug(WebSocketConstants.LOG_MESSAGE, "Couldn't connect to one of the server in the targets: ",
+                    targets);
+        }
+        return false;
+    }
+
+    /**
+     * Establishes connection with the endpoint.
+     *
+     * @param clientConnector -  the webSocket client connector
+     * @param webSocketClient - the WebSocket client
+     * @param wsService - the WebSocket service
+     */
+    public static void establishWebSocketConnection(WebSocketClientConnector clientConnector,
+                                                    ObjectValue webSocketClient, WebSocketService wsService) {
         boolean readyOnConnect = webSocketClient.getMapValue(CLIENT_ENDPOINT_CONFIG).getBooleanValue(
                 WebSocketConstants.CLIENT_READY_ON_CONNECT);
         ClientHandshakeFuture handshakeFuture = clientConnector.connect();
-        handshakeFuture.setWebSocketConnectorListener(clientConnectorListener);
         CountDownLatch countDownLatch = new CountDownLatch(1);
-        if (WebSocketUtil.hasRetryConfig(webSocketClient)) {
-            handshakeFuture.setClientHandshakeListener(new WebSocketClientHandshakeListenerForRetry(webSocketClient,
-                    wsService, clientConnectorListener, readyOnConnect, countDownLatch,
-                    (RetryContext) webSocketClient.getNativeData(WebSocketConstants.RETRY_CONFIG)));
+        setListenersToHandshakeFuture(handshakeFuture, webSocketClient, wsService, countDownLatch, readyOnConnect);
+        // Sets the countDown latch for every handshake
+        waitForHandshake(webSocketClient, countDownLatch, wsService);
+    }
+
+    /**
+     * Sets listeners to the handshake future.
+     *
+     * @param handshakeFuture - the handshake future
+     * @param webSocketClient - the WebSocket client
+     * @param wsService - the WebSocket service
+     * @param countDownLatch - the countdown latch
+     * @param readyOnConnect - the ready on connect
+     */
+    private static void setListenersToHandshakeFuture(ClientHandshakeFuture handshakeFuture,
+                                                      ObjectValue webSocketClient, WebSocketService wsService,
+                                                      CountDownLatch countDownLatch, boolean readyOnConnect) {
+        ConnectorListener connectorListener = (ConnectorListener) webSocketClient.getNativeData(
+                WebSocketConstants.CLIENT_LISTENER);
+        handshakeFuture.setWebSocketConnectorListener(connectorListener);
+        WebSocketHandshakeListener webSocketHandshakeListener = new WebSocketHandshakeListener(webSocketClient,
```

Comment:
This also we can move to respective Endpoints and pass it along inside WebSocketClient native data right.

Reply:
Every handshake needs to set the coundownLatch as it handles the handshake's timeout. therefore we can't move this to respective Endpoints.

Reply:
Yeah, can't we create handshake listener's along with the count down listeners in the respective init endpoint classes ?

Reply:
we can change the readyOnConnect value by calling the ready() and store the some connection properties into the webSocketClient and use it in the iteration connection. 
Therefore these values are also dynamic. So, we can't move this to respective Endpoints.